lib/common/rules.c

```diff
@@ -1004,100 +1004,153 @@ value_from_source(const char *value, enum pcmk__reference_source source,
         } else {
-            pcmk__config_err("Treating expression %s as not passing "
-                             "because '%s' is not a valid " PCMK_XA_OPERATION,
-                             pcmk__s(id, "without ID"), op);
+            pcmk__config_err("Treating " PCMK_XE_EXPRESSION " %s as not "
+                             "passing because '%s' is not a valid "
+                             PCMK_XA_OPERATION, id, op);
         }
-        return pcmk_rc_unpack_error;
+        rc = pcmk_rc_unpack_error;
+        goto done;
     }
 
-    if (rule_input != NULL) {
-        char *resolved_attr = NULL;
-        enum pcmk__reference_source source = pcmk__source_unknown;
-
-        // Expand any regular expression submatches (%0-%9) in attribute name
-        resolved_attr = pcmk__replace_submatches(attr, rule_input->rsc_id,
-                                                 rule_input->rsc_id_submatches,
-                                                 rule_input->rsc_id_nmatches);
-        if (resolved_attr != NULL) {
-            attr = (const char *) resolved_attr;
-            attr_allocated = TRUE;
-        }
+    // How reference value is obtained (literal, resource meta-attribute, etc.)
+    source_s = crm_element_value(expression, PCMK_XA_VALUE_SOURCE);
+    source = pcmk__parse_source(source_s);
+    if (source == pcmk__source_unknown) {
+        // Not possible with schema validation enabled
+        // @COMPAT Fail expression once we can break backward compatibility
+        pcmk__config_warn("Expression %s has invalid " PCMK_XA_VALUE_SOURCE
+                          " value '%s', using default "
+                          "('" PCMK_VALUE_LITERAL "')", id, source_s);
+        source = pcmk__source_literal;
+    }
 
-        // Get value appropriate to PCMK_XA_VALUE_SOURCE
-        source = pcmk__parse_source(value_source);
-        if (source == pcmk__source_unknown) {
-            // Not possible with schema validation enabled
-            // @COMPAT Fail expression once we can break backward compatibility
-            pcmk__config_warn("Expression %s has invalid " PCMK_XA_VALUE_SOURCE
-                              " value '%s', using default "
-                              "('" PCMK_VALUE_LITERAL "')",
-                              pcmk__s(id, "without ID"), value_source);
-            source = pcmk__source_literal;
-        }
-        value = value_from_source(value, source, rule_input);
+    // Get and validate reference value
+    value = crm_element_value(expression, PCMK_XA_VALUE);
+    switch (comparison) {
+        case pcmk__comparison_defined:
+        case pcmk__comparison_undefined:
+            if (value != NULL) {
+                pcmk__config_warn("Ignoring " PCMK_XA_VALUE " in "
+                                  PCMK_XE_EXPRESSION " %s because it is unused "
+                                  "when " PCMK_XA_BOOLEAN_OP " is %s", id, op);
+                rc = pcmk_rc_unpack_error;
```

Comment:
I don't think we want to return an error here; probably copy-paste error.

Reply:
definitely